#  Задание

Есть сервис для чтения книг по подписке. 
Первая задача  — проанализировать базу данных. 

Задания для запросов

•	Посчитайте, сколько книг вышло после 1 января 2000 года;

•	Для каждой книги посчитайте количество обзоров и среднюю оценку;

•	Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры;

•	Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;

•	Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок.



# Загрузка и предобработка

## Загрузка  необходимых библиотек

In [2]:
# импортируем библиотеки
import pandas as pd
from sqlalchemy import create_engine

In [3]:

# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
 db_config['pwd'],
 db_config['host'],
 db_config['port'],
 db_config['db'])
# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'}) 

## Запрос данных таблицы books

Таблица books
Содержит данные о книгах:
•	book_id — идентификатор книги;
•	author_id — идентификатор автора;
•	title — название книги;
•	num_pages — количество страниц;
•	publication_date — дата публикации книги;
•	publisher_id — идентификатор издателя.

In [4]:
# зададим функцию
def show_sql(query):
    return pd.io.sql.read_sql(query, con = engine,)

In [5]:
# выведем данные таблицы books
query = '''
           SELECT *   
           FROM books
           LIMIT 5
        ''' 
show_sql(query,)



book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

## Запрос данных таблицы authors

Таблица authors
Содержит данные об авторах:
•	author_id — идентификатор автора;
•	author — имя автора.

In [6]:
# выведем данные таблицы books
query = '''
           SELECT *   
           FROM authors
        ''' 
show_sql(query)



author_id                                author
0            1                            A.S. Byatt
1            2        Aesop/Laura Harris/Laura Gibbs
2            3                       Agatha Christie
3            4                         Alan Brennert
4            5              Alan Moore/David   Lloyd
..         ...                                   ...
631        632         William Strunk Jr./E.B. White
632        633                           Zadie Smith
633        634                 Zilpha Keatley Snyder
634        635                    Zora Neale Hurston
635        636  Åsne Seierstad/Ingrid Christopherson

[636 rows x 2 columns]

## Запрос данных таблицы publishers

Таблица publishers
Содержит данные об издательствах:
•	publisher_id — идентификатор издательства;
•	publisher — название издательства;

In [7]:
# выведем данные таблицы publishers
query = '''
           SELECT *   
           FROM publishers
        ''' 
show_sql(query)


publisher_id                          publisher
0               1                                Ace
1               2                           Ace Book
2               3                          Ace Books
3               4                      Ace Hardcover
4               5  Addison Wesley Publishing Company
..            ...                                ...
335           336         Workman Publishing Company
336           337                         Wyatt Book
337           338              Yale University Press
338           339                           Yearling
339           340                     Yearling Books

[340 rows x 2 columns]

## Запрос данных таблицы ratings

Таблица ratings
Содержит данные о пользовательских оценках книг:
•	rating_id — идентификатор оценки;
•	book_id — идентификатор книги;
•	username — имя пользователя, оставившего оценку;
•	rating — оценка книги.

In [8]:
# выведем данные таблицы ratings
query = '''
           SELECT *   
           FROM ratings
        ''' 
show_sql(query)


rating_id  book_id        username  rating
0             1        1      ryanfranco       4
1             2        1   grantpatricia       2
2             3        1    brandtandrea       5
3             4        2        lorichen       3
4             5        2     mariokeller       2
...         ...      ...             ...     ...
6451       6452     1000  carolrodriguez       4
6452       6453     1000         wendy18       4
6453       6454     1000      jarvispaul       5
6454       6455     1000           zross       2
6455       6456     1000         fharris       5

[6456 rows x 4 columns]

## Запрос данных таблицы reviews

Таблица reviews
Содержит данные о пользовательских обзорах на книги:
•	review_id — идентификатор обзора;
•	book_id — идентификатор книги;
•	username — имя пользователя, написавшего обзор;
•	text — текст обзора.

In [9]:
# выведем данные таблицы reviews
query = '''
           SELECT *   
           FROM reviews
        ''' 
show_sql(query)


review_id  book_id        username  \
0             1        1    brandtandrea   
1             2        1      ryanfranco   
2             3        2        lorichen   
3             4        3   johnsonamanda   
4             5        3     scotttamara   
...         ...      ...             ...   
2788       2789      999      martinadam   
2789       2790     1000         wknight   
2790       2791     1000  carolrodriguez   
2791       2792     1000         wendy18   
2792       2793     1000      jarvispaul   

                                                   text  
0     Mention society tell send professor analysis. ...  
1     Foot glass pretty audience hit themselves. Amo...  
2     Listen treat keep worry. Miss husband tax but ...  
3     Finally month interesting blue could nature cu...  
4     Nation purpose heavy give wait song will. List...  
...                                                 ...  
2788  Later hospital turn easy community. Fact same ...  
2789   Change lose answer close pressure. Spend so now.  
2790  Authority go who television entire hair guy po...  
2791  Or western offer wonder ask. More hear phone f...  
2792  Republican staff bit eat material measure plan...  

[2793 rows x 4 columns]

# Посчитаем, сколько книг вышло после 1 января 2000 года

In [10]:
# сколько книг вышло после 1 января 2000 года
query = '''
           SELECT COUNT(book_id) AS count_books  
           FROM books
           WHERE CAST (publication_date as date) >'01.01.2000'
        ''' 
show_sql(query)

count_books
0          819

# Для каждой книги посчитаем количество обзоров и среднюю оценку

In [11]:
# # Для каждой книги посчитаем количество обзоров и среднюю оценку
query = '''
           SELECT b.title,
               count_review,
               avg_rating              
               
           FROM 
           (SELECT book_id,
              AVG (rating)AS avg_rating
           FROM ratings           
           GROUP BY book_id)AS one
           LEFT JOIN (SELECT book_id,
               COUNT(review_id) AS count_review
           FROM reviews           
                     
           GROUP BY book_id) AS two ON one.book_id=two.book_id
           LEFT JOIN books AS b ON two.book_id=b.book_id
           ORDER BY two.count_review DESC         
           
        ''' 
show_sql(query)

title  count_review  \
0                                                 None           NaN   
1                                                 None           NaN   
2                                                 None           NaN   
3                                                 None           NaN   
4                                                 None           NaN   
..                                                 ...           ...   
995  Everyday Italian: 125 Simple and Delicious Rec...           1.0   
996  Dragons of Winter Night (Dragonlance: Chronicl...           1.0   
997                                         Lysistrata           1.0   
998                                  History of Beauty           1.0   
999                                 First Things First           1.0   

     avg_rating  
0      3.666667  
1      4.000000  
2      3.000000  
3      5.000000  
4      4.000000  
..          ...  
995    3.500000  
996    4.000000  
997    4.000000  
998    2.500000  
999    4.000000  

[1000 rows x 3 columns]

# Определим издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключим из анализа брошюры

In [13]:
#  выведем издательство
query = '''
           SELECT COUNT(book_id) AS books_count,
                  best_publisher.publisher
           FROM 
           (SELECT b.book_id,
                  pu.publisher
           FROM publishers as pu
           RIGHT OUTER JOIN  books AS b ON pu.publisher_id=b.publisher_id
           WHERE num_pages > 50) AS best_publisher
           GROUP BY publisher
           ORDER BY COUNT(book_id) DESC
           LIMIT 1
           

        ''' 
show_sql(query)

books_count      publisher
0           42  Penguin Books

# Определим автора с самой высокой средней оценкой книг — учитём только книги с 50 и более оценками

In [14]:
# выведем автора
query = '''

           SELECT author,
                 AVG(avg_rating)
           FROM
           (SELECT book_id,
                  COUNT(rating_id) AS count_rating_id,
                  AVG(rating) AS avg_rating
                 
           FROM ratings
           GROUP BY book_id
           ORDER BY count_rating_id DESC) AS b_r
           LEFT OUTER JOIN  books AS b ON b_r.book_id=b.book_id
           JOIN authors AS a ON b.author_id=a.author_id
           WHERE count_rating_id>50
           GROUP BY author
           ORDER BY AVG DESC
           LIMIT 1
        ''' 
show_sql(query)

author       avg
0  J.K. Rowling/Mary GrandPré  4.283844

#### Исправлено

# Посчитаем среднее количество обзоров от пользователей, которые поставили больше 50 оценок

In [15]:
#выведем среднее количество обзоров от пользователей
query = '''
           SELECT AVG(review_count) AS avg_review_count
           FROM
           (SELECT username,
                   COUNT(review_id) AS review_count
           FROM
           reviews 
           GROUP BY username) AS two 
           WHERE username IN (SELECT username
          FROM
           (SELECT username,
                  COUNT(rating_id) AS count_rating 
           FROM ratings
           GROUP BY username
           ORDER BY count_rating DESC
           LIMIT 6)AS one)
           
        '''
show_sql(query)

avg_review_count
0         24.333333

# Вывод

  В результате настоящего исследования были проанализированы базы данных с информацией о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут сформулировать ценностное предложение для нового продукта сервис для чтения книг по подписке. 
 В соответствии с заданием были загружены данные, сформированы необходимые запросы, получены следующие результаты:
 
 * результат запроса о количестве книг, вышедших после 1 января 2000 года - 819 единиц;
 
 * выведена таблица о данных количества обзоров и средней оценки  каждой книги ;
 
 * Определено издательство, которое выпустило наибольшее число книг толще 50 страниц — Penguin Books выпустило 42 книги;
 
 * Определён автора с самой высокой средней оценкой книг (учитены только книги с 50 и более оценками)- J.K. Rowling/Mary GrandPré, средний рейтинг её книг - 4.28;
 
 * Посчитано среднее количество обзоров от пользователей, которые поставили больше 50 оценок - 24.33.
 
 